tests/transaction.test.ts

```diff
@@ -0,0 +1,95 @@
+import { describe, test, assert } from "vitest";
+import { signal } from "../src/index";
+import { Transaction } from "../src/transaction";
+
+function getApp() {
+  class Obj {
+    @signal accessor value = 0;
+  }
+  const app = new Obj();
+  return app;
+}
+
+describe("transaction", () => {
+  test("rollback should work", () => {
+    const app = getApp();
+    app.value = 10;
+    const transaction = new Transaction();
+    transaction.execute(() => {
+      app.value = 20;
+    });
+    assert.equal(app.value, 20);
+    transaction.rollback();
+    assert.equal(app.value, 10);
+  });
+  test("commit should work", () => {
+    const app = getApp();
+    app.value = 10;
+    const transaction = new Transaction();
+    transaction.execute(() => {
+      app.value = 20;
+    });
+    assert.equal(app.value, 20);
+    transaction.commit();
+    assert.equal(app.value, 20);
+  });
+  test("should work with nested transactions", () => {
+    const app = getApp();
+    app.value = 10;
+    const transaction = new Transaction();
+    transaction.execute(() => {
+      app.value = 20;
+      const nestedTransaction = new Transaction();
+      nestedTransaction.execute(() => {
+        app.value = 30;
+      });
+      assert.equal(app.value, 30);
+      nestedTransaction.rollback();
+      assert.equal(app.value, 20);
+    });
+    assert.equal(app.value, 20);
+    transaction.rollback();
+    assert.equal(app.value, 10);
+  });
+  test("should execute mutation in constructor", () => {
+    const app = getApp();
+    const transaction = new Transaction(() => {
+      app.value = 20;
+    });
+    assert.equal(app.value, 20);
+    transaction.rollback();
+    assert.equal(app.value, 0);
+  });
+  test("should work with promises", async () => {
+    const app = getApp();
+    app.value = 10;
+    const transaction = new Transaction(() => {
+      app.value = 20;
+    });
+    const asyncRequest = new Promise((resolve) => {
+      setTimeout(() => {
+        resolve(true);
+      }, 100);
+    });
+    await transaction.follow(asyncRequest);
```

Comment:
ooo this is interesting! transactions can extend to promises!

Reply:
yep, but I have one doubts here, likely we need to wait for all promise.then to be completed before applying transaction commit, because some logic likely already wrapped to original resolve promise